// w3c-webdriver/lib/rest.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DELETE = exports.POST = exports.GET = undefined;

var _http = require('http');

var _http2 = _interopRequireDefault(_http);

var _url = require('url');

var _url2 = _interopRequireDefault(_url);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function findError({ status, value }) {
  if (!status && (!value || !value.error)) {
    return null;
  }

  const message = value.message,
        error = value.error;


  return new Error(`WebDriverError(${error || status}): ${message}`);
}

function sendRequest(method, url, body) {
  const jsonBody = JSON.stringify(body);
  const urlParts = _url2.default.parse(url);
  const options = {
    hostname: urlParts.hostname,
    port: urlParts.port,
    path: urlParts.path,
    method,
    headers: {}
  };

  if (body) {
    options.headers['Content-Length'] = Buffer.byteLength(jsonBody);
    options.headers['Content-Type'] = 'application/json';
  }

  return new Promise((resolve, reject) => {
    const request = _http2.default.request(options, response => {
      const chunks = [];
      response.setEncoding('utf8');
      response.on('data', chunk => {
        chunks.push(chunk);
      });
      response.on('end', () => {
        let responseBody;

        try {
          responseBody = JSON.parse(chunks.join(''));
        } catch (err) {
          reject(err);
        }

        const error = findError(responseBody);

        if (error) {
          reject(error);
          return;
        }

        resolve(responseBody);
      });
    });

    request.on('error', error => reject(error));

    if (body) {
      request.write(jsonBody);
    }
    request.end();
  });
}

const GET = exports.GET = url => sendRequest('GET', url);
const POST = exports.POST = (url, body) => sendRequest('POST', url, body);
const DELETE = exports.DELETE = (url, body) => sendRequest('DELETE', url, body);
